---
title: "Intro to Transcriptomics"
author: Nicole Black, Wade Boohar, Kayla Xu
date: 07/17/22
updated: 09/26/23
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on BB.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partnerâ€™s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In Transcriptomics_Tutorial.Rmd, you learned how to manipulate the SummarizedExperiment data structure, turn it into more readable dataframes, complete basic analysis using those dataframes, and performed a differential expression analysis based on age. Before saving the rna_counts, rna_clinical, and rna_genes however, we masked out any sample that was not a tumor. In this assignment, you will need to reinstantiate the rna_se object and then perform differential expression analysis based on tumor status.

*Pre-Assignment*
Use knitr function to set your working directory to your analysis_data folder in qbio_490_name.
```{r}
    knitr::opts_knit$set(root.dir = normalizePath("/Users/kevinliu/desktop/qbio_490/QBIO_490_kevinliu/analysis_data"))
```

Load all necessary packages and instantiate the rna_se object
```{r}
# install, only if needed, and load packages (SummarizedExperiment and TCGAbiolinks)
if (!require("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install(version = "3.17")
if (!require("TCGAbiolinks", quietly = TRUE))
BiocManager::install("TCGAbiolinks")
if (!require("maftools", quietly = TRUE))
BiocManager::install("maftools")
if (!require("DESeq2", quietly = TRUE))
BiocManager::install("DESeq2")
if (!require("EnhancedVolcano", quietly = TRUE))
BiocManager::install("EnhancedVolcano")
library(BiocManager)
library(TCGAbiolinks)
library(maftools)
library(ggplot2)
library (SummarizedExperiment)
library(DESeq2)
library(EnhancedVolcano)


rna_query <- GDCquery(project ="TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
#GDCdownload(rna_query)

rna_se <- GDCprepare(rna_query)
```


*1*
Perform all data cleaning and label manipulation steps (hint: applying colnames and rownames to make the dataframes more readable) except for masking out the patient with an NA for age and the "Solid Tissue Normal" sample types.
```{r}
rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)

treatments_mask <- ifelse(colnames(rna_clinical) == "treatments", F, T)
rna_clinical <- rna_clinical[, treatments_mask]

primary_mask <- ifelse(colnames(rna_clinical) == "primary_site", F, T)
rna_clinical <- rna_clinical[, primary_mask]

diseases_mask <- ifelse(colnames(rna_clinical) == "disease_type", F, T)
rna_clinical <- rna_clinical[, diseases_mask]

rna_genes <- rna_se@rowRanges@elementMetadata

rna_genes <- as.data.frame(rna_genes)

rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)

rna_clinical$age_category <- ifelse(rna_clinical$age_at_index <= 58, "young", "old")

age_NA_mask <- !is.na(rna_clinical$age_category)
#removing NAs
rna_clinical <- rna_clinical[age_NA_mask, ]

rna_counts <- rna_counts[, age_NA_mask]

rownames(rna_genes) <- rna_genes$gene_id

rownames(rna_counts) <- rna_genes$gene_id
colnames(rna_counts) <- rownames(rna_clinical)


```


*2*
<<<<<<< HEAD
Determine and define your experimental design for the differential expression analysis. Control for covariance of age, gender, OR vital_status. Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what data type are they by default? do you need to handle unknown values?)
=======

>>>>>>> 75150ea0bc4dd3d410d7944d4b31b80ff89c3865

```{r}
# rna_clinical$prior_malignancy
# rna_clinical$vital_status
# rna_clinical$gender

na_mask_DESeq2 <- !is.na(rna_clinical$prior_malignancy)
rna_clinical_DESeq2 <- rna_clinical[na_mask_DESeq2, ]

na_mask_DESeq2 <- !is.na(rna_clinical$gender)
rna_clinical_DESeq2 <- rna_clinical[na_mask_DESeq2, ]
rna_clinical_DESeq2$gender <- factor(rna_clinical_DESeq2$gender)

na_mask_DESeq2 <- !is.na(rna_clinical$vital_status)
rna_clinical_DESeq2 <- rna_clinical[na_mask_DESeq2, ]
rna_clinical_DESeq2$vital_status <- factor(rna_clinical_DESeq2$vital_status)

row_sums <- rowSums(rna_counts)

low_counts_mask <- ifelse(row_sums < 10, F, T)

rna_counts_cleaned <- rna_counts[low_counts_mask, ]

rna_genes_cleaned <- rna_genes[low_counts_mask, ]

```


*3*
Perform the differential expression analysis, create a volcano plot, and interpret what genes from each of the 6 sections means (assuming thresholds of -log10padj = 2 and log2FoldChange = 1)
```{r}
rna_clinical_DESeq2$prior_malignancy <- factor(rna_clinical_DESeq2$prior_malignancy)

?DESeqDataSetFromMatrix
dds <- DESeqDataSetFromMatrix(countData = rna_counts_cleaned,
                             colData = rna_clinical_DESeq2,
                             design = ~ vital_status + gender + prior_malignancy)


?DESeq
dds_obj <- DESeq(dds) # note: this will likely take a long time (ie 45 minutes to 2 hours)

?resultsNames
resultsNames(dds_obj)  # see what comparisons got run
head(results)
# get the yes vs. no comparison
?results
results <- results(dds_obj, format = "DataFrame", contrast = c("prior_malignancy", "yes", "no")) 

results <- data.frame(rna_genes_cleaned$gene_name, results@rownames, results@listData$log2FoldChange, results@listData$pvalue, results@listData$padj, -log10(results@listData$padj))


colnames(results) <- c("gene_name", "gene_id", "log2FoldChange", "pvalue", "padj", "-log10(padj)") ## FIX column names

# add gene_id as the row names for results
row.names(results) <- rna_genes_cleaned$gene_id

par(mar=c(0.5,1,1,0.5))
EnhancedVolcano(results, 
                lab = results$gene_name, 
                x = 'log2FoldChange', 
                y = 'pvalue',
                titleLabSize = 10,
                subtitleLabSize = 6,
                captionLabSize = 5,
                axisLabSize = 10,
                legendPosition = "right",
                legendLabSize = 10,
                labSize = 3,
                pointSize = 1,
                gridlines.major = FALSE,
                gridlines.minor = FALSE,
                #shape = c(1, 4, 23, 25),
)


```

top-right genes: These are significant upregulated genes
bottom-right genes: These are upregulated but not significant
top-left genes: These are significant down regulated genes
bottom-left genes: These are down regulated genes but not significant 
top-middle genes: These are significant genes but meet neither thresholds for upregulated and down regulated
bottom-middle genes: These are unsignificant genes that meet neither thresholds for upregulated and down regulated

Save the picture of the volcano plot (using either ggsave() or right clicking and manually downloading the image and push this .Rmd and the image to GitHub)
