---
title: "QBIO_final_project_hpv"
output: html_document
date: "2023-11-30"
---

```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("/Users/kevinliu/Desktop/QBIO_490/Final Project/qbio_tcga_hnsc/analysis_data"))
```


```{r}
# install, only if needed, and load packages (SummarizedExperiment and TCGAbiolinks)
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.17")
if (!require("TCGAbiolinks", quietly = TRUE))
  BiocManager::install("TCGAbiolinks")
if (!require("maftools", quietly = TRUE))
  BiocManager::install("maftools")
if (!require("DESeq2", quietly = TRUE))
  BiocManager::install("DESeq2")
if (!require("EnhancedVolcano", quietly = TRUE))
  BiocManager::install("EnhancedVolcano")
if (!require(survival)) {
  install.packages("survival")}
if (!require(survminer)) {
  install.packages("survminer")}
if (!require(mltools)) {
  install.packages("dplyr")}

library(BiocManager)
library(TCGAbiolinks)
library(maftools)
library(ggplot2)
library (SummarizedExperiment)
library(DESeq2)
library(EnhancedVolcano)
library(survminer)
library(survival)
library(dplyr)

```

```{r}
#neccessary downloads and preparation for clinical files

clin_query <- GDCquery(project = "TCGA-HNSC",
                       data.category = "Clinical",
                       data.type = "Clinical Supplement",
                       data.format = 'BCR Biotab')
GDCdownload(clin_query)
clinical.BCRtab.all <- GDCprepare(clin_query)  
clinic <- clinical.BCRtab.all$clinical_patient_hnsc[-c(1,2),]

colnames(clinic)[ colnames(clinic) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"

write.csv(clinic, "/Users/kevinliu/desktop/qbio_490/QBIO_490_kevinliu/analysis_data/hnsc_clinical_data.csv", row.names=FALSE)

clinic <- read.csv("/Users/kevinliu/desktop/qbio_490/QBIO_490_kevinliu/analysis_data/hnsc_clinical_data.csv")


```

```{r}
#neccessary downloads and preparation for rna_se files

rna_query <- GDCquery(project ="TCGA-HNSC",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)

rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)
rna_clinical$Tumor_Sample_Barcode <- rna_clinical$patient

```
```{r}
#neccessary downloads and preparation for MAF files

clinical <- read.csv("/Users/kevinliu/desktop/qbio_490/QBIO_490_kevinliu/analysis_data/hnsc_clinical_data.csv")
colnames(clinical)[ colnames(clinical) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"
write.csv(clinical, "/Users/kevinliu/desktop/qbio_490/QBIO_490_kevinliu/analysis_data/hnsc_clinical_data.csv", row.names = FALSE)

maf_query <- GDCquery(
  project = "TCGA-HNSC", 
  data.category = "Simple Nucleotide Variation", 
  access = "open", # we only have access to somatic mutations which are open access
  data.type = "Masked Somatic Mutation", 
  workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking"
)

GDCdownload(maf_query)

maf <- GDCprepare(maf_query) # as long as it runs, ignore any errors

maf_object <- read.maf(maf = maf, 
                       clinicalData = rna_clinical,
                       isTCGA = TRUE)
```



Investigate the different groups of smokers/drinkers
```{r}
# rna_clinical 566 x 84

mask_alc <- ifelse(rna_clinical$alcohol_history == "Yes", TRUE, FALSE)
mask_noalc <- ifelse(rna_clinical$alcohol_history == "No", TRUE, FALSE)

both <- rna_clinical[mask_alc, ]
both <- both[!is.na(both$cigarettes_per_day) | 
               !is.na(both$years_smoked), ]

alc_only <- rna_clinical[mask_alc, ]
alc_only <- alc_only[is.na(alc_only$cigarettes_per_day) & is.na(alc_only$years_smoked), ]

cig_only <- rna_clinical[mask_noalc, ]
cig_only <- cig_only[(!is.na(cig_only$cigarettes_per_day) |
                       !is.na(cig_only$years_smoked)), ]

# DF containing pt that does neither
none <- rna_clinical[mask_noalc, ]
none <- none[is.na(none$cigarettes_per_day) & is.na(none$years_smoked), ]

# Any patients not in None df
any <- rna_clinical[!(rna_clinical$patient %in% none$patient), ]

missing <- any[!(any$patient %in% alc_only$patient) &
                 !(any$patient %in% cig_only$patient) &
                 !(any$patient %in% both$patient), ]

# Actual dataset to use after removing the "Not Reported" group
any_cleaned <- any[!(any$patient %in% missing$patient),]
clinical_cleaned <- rna_clinical[!(rna_clinical$patient %in% missing$patient), ]
none_cleaned <- clinical_cleaned[!(clinical_cleaned$patient %in% any_cleaned$patient), ]

# Usage column to classify patients into 4 categories
clinical_cleaned$usage <- ifelse(clinical_cleaned$patient %in% both$patient, "Both", 
                             ifelse(clinical_cleaned$patient %in% alc_only$patient, "Alcohol", 
                                    ifelse(clinical_cleaned$patient %in% 
                                             cig_only$patient, "Cigarettes", "None")))

clinical_cleaned$split <- ifelse(clinical_cleaned$patient %in% any_cleaned$patient, "Any", "None")

rna_clinical <- clinical_cleaned

```



HPV groups (clinic)
```{r}
negative_mask <- ifelse(clinical$hpv_status_p16 == "Negative", TRUE, FALSE)
positive_mask <- ifelse(clinical$hpv_status_p16 == "Positive", TRUE, FALSE)

negative_only <- clinical[negative_mask, ]
positive_only <- clinical[positive_mask, ]

clinical$negative_only <- ifelse(clinical$Tumor_Sample_Barcode %in% negative_only$Tumor_Sample_Barcode, "Yes", "No")
clinical$positive_only <- ifelse(clinical$Tumor_Sample_Barcode %in% positive_only$Tumor_Sample_Barcode, "Yes", "No")

```

CoOncoplots
```{r}
#variables for plotting

maf_object@clinical.data$negative_only <- ifelse(maf_object@clinical.data$patient %in% negative_only$Tumor_Sample_Barcode, "Yes", "No")

maf_object@clinical.data$positive_only <- ifelse(maf_object@clinical.data$patient %in% positive_only$Tumor_Sample_Barcode, "Yes", "No")



negative_only_barcodes <- maf_object@clinical.data[ifelse(maf_object@clinical.data$negative_only == "Yes", TRUE, FALSE), ]$Tumor_Sample_Barcode
negative <- subsetMaf(maf = maf_object,
                        tsb = negative_only_barcodes)

positive_only_barcodes <- maf_object@clinical.data[ifelse(maf_object@clinical.data$positive_only == "Yes", TRUE, FALSE), ]$Tumor_Sample_Barcode
positive <- subsetMaf(maf = maf_object,
                           tsb = positive_only_barcodes)

coOncoplot(m1 = negative,
           m2 = positive,
           m1Name = "negative",
           m2Name = "positive",
           anno_height = 4,
           outer_mar = 4,
           barcode_mar = 1.5,
           geneNamefont = 0.7,
           gene_mar = 1.9,
           titleFontSize = 0.8,
           borderCol = NA)

```

More oncoplots comparing subsetted hpv and substance groups
```{r}
# #Prepare variables for plotting
# mask_alc <- ifelse(maf_object@clinical.data$alcohol_history == "Yes", TRUE, FALSE)
# mask_noalc <- ifelse(maf_object@clinical.data$alcohol_history == "No", TRUE, FALSE)
# 
# both <- maf_object@clinical.data[mask_alc, ]
# both <- both[(ifelse(both$cigarettes_per_day != "NA", TRUE, FALSE) |
#                        ifelse(both$years_smoked != "NA", TRUE, FALSE)), ]
# 
# alc_only <- as.data.frame(maf_object@clinical.data[mask_alc, ])
# alc_only <- alc_only[(ifelse(alc_only$cigarettes_per_day == "NA", TRUE, FALSE) & ifelse(alc_only$years_smoked == "NA", TRUE, FALSE)), ]
# 
# cig_only <- as.data.frame(maf_object@clinical.data[mask_noalc, ])
# cig_only <- cig_only[(ifelse(cig_only$cigarettes_per_day != "NA", TRUE, FALSE) |
#                        ifelse(cig_only$years_smoked != "NA", TRUE, FALSE)), ]
# 
# # DF containing pt that does neither
# none <- as.data.frame(maf_object@clinical.data[mask_noalc, ])
# none <- none[(ifelse(none$cigarettes_per_day == "NA", TRUE, FALSE) & ifelse(none$years_smoked == "NA", TRUE, FALSE)), ]
# 
# # Any patients not in None df
# any <- as.data.frame(maf_object@clinical.data[!(as.data.frame(maf_object@clinical.data$patient) %in% none$patient), ])
# 
# missing <- any[!(any$patient %in% alc_only$patient) &
#                  !(any$patient %in% cig_only$patient) &
#                  !(any$patient %in% both$patient), ]
# 
# # Actual dataset to use after removing the "Not Reported" group
# any_cleaned <- any[!(any$patient %in% missing$patient),]
# clinical_cleaned <- as.data.frame(maf_object@clinical.data[!(as.data.frame(maf_object@clinical.data)$patient %in% missing$patient), ])
# none_cleaned <- clinical_cleaned[!(clinical_cleaned$patient %in% any_cleaned$patient), ]
# 
# # Usage column to classify patients into 4 categories
# clinical_cleaned$usage <- ifelse(clinical_cleaned$patient %in% both$patient, "Both", 
#                              ifelse(clinical_cleaned$patient %in% alc_only$patient, "Alcohol", 
#                                     ifelse(clinical_cleaned$patient %in% 
#                                              cig_only$patient, "Cigarettes", "None")))
# 
# clinical_cleaned$split <- ifelse(clinical_cleaned$patient %in% any_cleaned$patient, "Any", "None")
# 
# maf_object@clinical.data <- clinical_cleaned




both_barcodes <- maf_object@clinical.data[ifelse(maf_object@clinical.data$usage == "Both", TRUE, FALSE), ]$Tumor_Sample_Barcode
both <- subsetMaf(maf = maf_object,
                        tsb = both_barcodes)

none_barcodes <- maf_object@clinical.data[ifelse(maf_object@clinical.data$only_cig == "Yes", TRUE, FALSE), ]$Tumor_Sample_Barcode
none <- subsetMaf(maf = maf_object,
                           tsb = none_barcodes)

coOncoplot(m1 = only_alc,
           m2 = only_cig,
           m1Name = "only_alc",
           m2Name = "only_cig",
           anno_height = 4,
           outer_mar = 4,
           barcode_mar = 1.5,
           geneNamefont = 0.8,
           gene_mar = 1.9,
           titleFontSize = 0.8,
           borderCol = NA)

```

Somatic Interactions
```{r}
somaticInteractions(maf = negative,
                    genes = NULL,
                    pvalue = c(0.05, 0.1)) ## these are the default p-value significance levels

somaticInteractions(maf = positive,
                    genes = NULL,
                    pvalue = c(0.05, 0.1)) ## these are the default p-value significance levels
```

Lollipop
```{r}
lollipopPlot2(m1 = negative, 
              m2 = positive, 
              m1_name = "negative",
              m2_name = "positive",
              gene = "TP53",
              showDomainLabel = FALSE
              )

lollipopPlot2(m1 = negative, 
              m2 = positive, 
              m1_name = "negative",
              m2_name = "positive",
              gene = "FAT1",
              showDomainLabel = FALSE
              )

lollipopPlot2(m1 = negative, 
              m2 = positive, 
              m1_name = "negative",
              m2_name = "positive",
              gene = "EP300",
              showDomainLabel = FALSE
              )

lollipopPlot2(m1 = negative, 
              m2 = positive, 
              m1_name = "negative",
              m2_name = "positive",
              gene = "ZNF750",
              showDomainLabel = FALSE
              )


```

KM Survival plot
```{r}

rna_clinical$negative_only <- ifelse(rna_clinical$patient %in% negative_only$Tumor_Sample_Barcode, "Yes", "No")

rna_clinical$positive_only <- ifelse(rna_clinical$patient %in% positive_only$Tumor_Sample_Barcode, "Yes", "No")

rna_clinical$survival_time <- ifelse(!is.na(rna_clinical$days_to_last_follow_up), rna_clinical$days_to_last_follow_up, rna_clinical$days_to_death)

rna_clinical$survival_time <- as.numeric(rna_clinical$survival_time)
rna_clinical <- rna_clinical[!is.na(rna_clinical$survival_time),]

investigate <- rna_clinical[ifelse(rna_clinical$negative_only =="Yes", TRUE, ifelse(rna_clinical$positive_only == "Yes", TRUE, FALSE)),]

investigate$hpv_status <- ifelse(investigate$negative_only == "Yes", "negative", "positive")

investigate$death_event <- ifelse(investigate$vital_status == "Dead", T, F)

investigate_positive <- investigate[ifelse(investigate$hpv_status == "positive", TRUE, FALSE), ]
  
investigate_negative <- investigate[ifelse(investigate$hpv_status == "negative", TRUE, FALSE), ]

#Prepare KM Plot variables using hpv negative vs hpv positive
survival_object <- Surv(time = investigate$survival_time,
                        event = investigate$death_event)

fit_object <- survfit(survival_object ~ hpv_status, data = investigate)

survplot <- ggsurvplot(fit_object,
                       pval=TRUE,
                       ggtheme = theme(plot.margin = unit(c(1,1,1,1), "cm")),
                       legend = 'right')

KM_plot <- survplot$plot + theme_bw() + theme(axis.title = element_text(size=20), 
                                              axis.text = element_text(size=16),
                                              legend.title = element_text(size=14),
                                              legend.text = element_text(size=12))

KM_plot

#Prepare KM Plot variables using different usage for hpv negative
survival_object <- Surv(time = investigate_negative$survival_time,
                        event = investigate_negative$death_event)

fit_object <- survfit(survival_object ~ usage, data = investigate_negative)

survplot <- ggsurvplot(fit_object,
                       pval=TRUE,
                       ggtheme = theme(plot.margin = unit(c(1,1,1,1), "cm")),
                       legend = 'right')

KM_plot <- survplot$plot + theme_bw() + theme(axis.title = element_text(size=20), 
                                              axis.text = element_text(size=16),
                                              legend.title = element_text(size=14),
                                              legend.text = element_text(size=12))

KM_plot

#Prepare KM Plot variables using different usage for hpv positive
survival_object <- Surv(time = investigate_positive$survival_time,
                        event = investigate_positive$death_event)

fit_object <- survfit(survival_object ~ usage, data = investigate_positive)

survplot <- ggsurvplot(fit_object,
                       pval=TRUE,
                       ggtheme = theme(plot.margin = unit(c(1,1,1,1), "cm")),
                       legend = 'right')

KM_plot <- survplot$plot + theme_bw() + theme(axis.title = element_text(size=20), 
                                              axis.text = element_text(size=16),
                                              legend.title = element_text(size=14),
                                              legend.text = element_text(size=12))

KM_plot


```

Table for hpv status and usage status
```{r}
addmargins(table(investigate$usage, investigate$hpv_status))

prop.table(table(investigate$usage, investigate$hpv_status))
```





